java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthVector.java

```diff
@@ -1250,8 +1251,10 @@ protected final void handleSafe(int index, int dataLength) {
     while (index >= getValueCapacity()) {
       reallocValidityAndOffsetBuffers();
     }
-    final int startOffset = lastSet < 0 ? 0 : getStartOffset(lastSet + 1);
-    while (valueBuffer.capacity() < (startOffset + dataLength)) {
+    final long startOffset = lastSet < 0 ? 0 : getStartOffset(lastSet + 1);
+    final long targetCapacity = startOffset + dataLength;
+    checkDataBufferSize(targetCapacity);
+    while (valueBuffer.capacity() < targetCapacity) {
       reallocDataBuffer();
     }
```

Comment:
Slightly unrelated, but why is this using a `while` loop? Ideally it would be more efficient to write:
```suggestion
    reallocDataBuffer(targetCapacity);
```

Reply:
I think the reallocation may not meet the memory request if the `dataLength` is larger than two times of `valueBuffer.capacity()`.  That I think the while loop is needed here.

Reply:
Yes. My point is that it's wasteful to reallocate several times in a row, instead of reallocating directly to the desired target capacity. Anyway, this was already the case before.

Reply:
Got your point. And updated the code. It should be the right way.